pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -431,17 +431,41 @@ func (a *Actuator) chooseHost(ctx context.Context, machine *machinev1beta1.Machi
 
 	availableHosts := []*bmh.BareMetalHost{}
 	for i, host := range hosts.Items {
-		if host.Available() {
-			if labelSelector.Matches(labels.Set(host.ObjectMeta.Labels)) {
-				log.Printf("Host '%s' matched hostSelector for Machine '%s'", host.Name, machine.Name)
-				availableHosts = append(availableHosts, &hosts.Items[i])
-			} else {
-				log.Printf("Host '%s' did not match hostSelector for Machine '%s'", host.Name, machine.Name)
-			}
-		} else if host.Spec.ConsumerRef != nil && consumerRefMatches(host.Spec.ConsumerRef, machine) {
+
+		if consumerRefMatches(host.Spec.ConsumerRef, machine) {
+			// if we see a host that thinks this machine is consuming
+			// it, we should oblige it
 			log.Printf("found host %s with existing ConsumerRef", host.Name)
 			return &hosts.Items[i], nil
 		}
+
+		if host.Spec.ConsumerRef != nil {
+			// consumerRef is set and does not match, so the host is
+			// in use
+			continue
+		}
+		if host.GetDeletionTimestamp() != nil {
+			// the host is being deleted
+			continue
+		}
+		if host.Status.ErrorMessage != "" {
+			// the host has some sort of error
+			continue
+		}
+		if host.Spec.ExternallyProvisioned {
+			// the host was provisioned by something else, we should
+			// not overwrite it
+			continue
```

Comment:
Actually, shouldn't we go further and say we are only interested in Hosts in the Ready provisioning state?

Reply:
Sure, I added that. I left the explicit checks for externally provisioned and errors in place for clarity.